### PR TITLE
Add test for ANA support

### PR DESCRIPTION
**Pull request by hreinecke**
This small patchset adds a test for ANA support.
In doing so I also had to update the loop transport to enable more than one port.
But with that the issues mentioned in https://github.com/osandov/blktests/pull/113 are gone.

### CODE REVIEW

**Comment by kawasaki** [comment]
I tried to run tests with the changes on kernel version 6.2, and found that number of nvme test cases fail. For example, nvme/004 fail after the commit "nvme: enable ANA support".

```
nvme/004 (test nvme and nvmet UUID NS descriptors)           [failed]
    runtime  1.776s  ...  0.598s
    --- tests/nvme/004.out      2023-02-28 19:15:13.425727236 +0900
    +++ /home/shin/kts/kernel-test-suite/src/blktests/results/nodev/nvme/004.out.bad    2023-03-03 09:12:21.674526288 +0900
    @@ -1,5 +1,5 @@
     Running nvme/004
    -91fdba0d-f87b-4c25-b80f-db7be1418b9e
    -uuid.91fdba0d-f87b-4c25-b80f-db7be1418b9e
    +323d5abb-ff8e-4a12-99ad-128e15baef41
    +uuid.323d5abb-ff8e-4a12-99ad-128e15baef41
     NQN:blktests-subsystem-1 disconnected 1 controller(s)
     Test complete
```

**Comment by kawasaki** [comment]
As for the commit message of the second patch, it would be helpful to note that ANA is the acronym of Asymmetric Namespace Access that NVMe spe.c 1.4 introduced. (If I'm wrong, please correct me)

**Review by kawasaki — CHANGES_REQUESTED** [review]
@hreinecke Thanks for the work. I reviewed and did test run, then made some feedbacks.

One more nit comment: the first three patches have various commit title prefixes "nvme", "nvme/rc" or "test/nvme". but they all modify test/nvme/rc file. The past commits do not have consistent prefix, but I think "nvme/rc" is the good choice.

File: tests/nvme/rc [hunk 1]
```diff
@@ -374,9 +372,33 @@ _create_nvmet_port() {
...
+	local portcfs
 	for ((port = 0; ; port++)); do
-		if [[ ! -e "${NVMET_CFS}/ports/${port}" ]]; then
+		portcfs="${NVMET_CFS}/ports/${port}"
+		if [[ ! -e "${portcfs}" ]]; then
 			break
+		else
+			tmp_trtype=$(cat "${portcfs}/addr_trtype")
+			tmp_traddr=$(cat "${portcfs}/addr_traddr")
+			tmp_adrfam=$(cat "${portcfs}/addr_adrfam")
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply tests/nvme/rc:383]: Well, not really. 'adrfam' can be derived directly from the transport address itself (as each address family has it's own transport address format). So a string comparison between transport addresses is sufficient here, and we can remove the 'adrfam' line.

File: tests/nvme/rc [hunk 2]
```diff
@@ -62,6 +62,10 @@ _nvme_requires() {
 		fc)
 			def_adrfam="fc"
 			;;
+		loop)
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply tests/nvme/rc]: Yes, will be doing so.

**Comment by kawasaki** [comment]
@hreinecke Thanks for reworking the patches. Today, I ran the test case nvme/047 and observed it fails.

```
nvme/047 (test ANA optimized/transitioning/inaccessible support) [failed]
    runtime  44.260s  ...  45.509s
    --- tests/nvme/047.out      2023-03-07 10:07:13.131098972 +0900
    +++ /home/shin/kts/kernel-test-suite/src/blktests/results/nodev/nvme/047.out.bad    2023-03-07 13:54:20.111062268 +0900
    @@ -1,28 +1,71 @@
     Running nvme/047
     ANA state failover
    -1: grpid 1 state inaccessible
    -1: grpid 2 state optimized
    -2: grpid 1 state optimized
    -2: grpid 2 state inaccessible
    +cat: /sys/class/nvme/nvme7/nvme7n1/ana_grpid: No such file or directory
    ...
    (Run 'diff -u tests/nvme/047.out /home/shin/kts/kernel-test-suite/src/blktests/results/nodev/nvme/047.out.bad' to see the entire diff)
```

It looks that the sysfs attribute file ana_grpid is not found under /sys/class/nvme/nvmeX/nvmeXn1. I wonder what I'm missing to make the test case pass. I use kernel v6.2 and nvme-cli v2.2.1. Do I need any changes on top of these?

One nit comment is shellcheck warning.  The change below for the last patch will avoid it.

```
diff --git a/tests/nvme/047 b/tests/nvme/047
index 0853d76..401f06e 100755
--- a/tests/nvme/047
+++ b/tests/nvme/047
@@ -58,7 +58,7 @@ execute_ana_test() {
        local port1=$3
        local port2=$4
        local mount_dir="/mnt/blktests"
-       local nproc=$(nproc)
+       local nproc
 
        switch_nvmet_anagroup "${port1}" "${port2}" "failover"
 
@@ -71,6 +71,7 @@ execute_ana_test() {
 
        _xfs_mkfs_and_mount "/dev/${nvmedev}n2" "${mount_dir}/ns2" >> "${FULL}" 2>&1
    
+       nproc=$(nproc)
        if (( nproc > 8 )); then
                nproc=8
        fi
```

**Comment by hreinecke** [comment]
And we can now retire this in favour of https://github.com/osandov/blktests/pull/146

**Pull request closed without merging**